### PR TITLE
Yet another SSRF query for Javascript

### PR DESCRIPTION
**Pull request by valeria-meli**
This query finds SSRFs (Server Side Request Forgeries). This is intended to be an alternative to the current Request Forgery query.

### Why a new query?
We use codeql queries at Mercado Libre to block pull requests. We need them to be very precise. 
We recommend input validation, not URL escaping, when fixing a SSRF. Some NGINx based routing tend to decode before routing, making our URL escape useless. But input validation always works.
So we adapted the original query to this blocking context and to accept different forms of input validation.

### CODE REVIEW

**Comment by intrigus-lgtm** [comment]
Please auto-format your (QL) code.
You can do this by running `codeql query format --in-place` on the changed files or by using the CodeQL extension for VS Code (I think it's called **Format Document** in VS Code).

File: javascript/ql/src/experimental/Security/CWE-918/SSRF.qll [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+/** TODO add comment */
+class TernaryOperatorSanitizerGuard extends TaintTracking::SanitizerGuardNode {
+    TaintTracking::SanitizerGuardNode originalGuard;
+
+    TernaryOperatorSanitizerGuard() {
+    exists(DataFlow::Node falseNode |
+        this.getAPredecessor+() = falseNode and 
+        falseNode.asExpr().(BooleanLiteral).mayHaveBooleanValue(false)
+    ) and
```
Comment (intrigus-lgtm) [inline javascript/ql/src/experimental/Security/CWE-918/SSRF.qll:36]: You should be able to simplify this to `this.getAPredecessor+().asExpr().(BooleanLiteral).mayHaveBooleanValue(false)`, I think.

**Review by asgerf — COMMENTED** [review]
👍  Thanks for the submission. It's very valuable to get feedback on our queries like this, and in time we'll most likely incorporate the new sanitizers into the standard library.

For now, please see my inline comment regarding the heuristic that was supposedly removed. I'd like to get clarity on what happened here and that the query works as described.

File: javascript/ql/src/experimental/Security/CWE-918/SSRF.qll [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  override predicate isSink(DataFlow::Node sink) { sink instanceof RequestForgery::Sink }
+
+  override predicate isSanitizer(DataFlow::Node node) {
+    super.isSanitizer(node) or
+    node instanceof RequestForgery::Sanitizer
+  }
+
+  override predicate isSanitizerEdge(DataFlow::Node source, DataFlow::Node sink) {
+    sanitizingPrefixEdge(source, sink)
+  }
```
[the remark this answers is not in the document]
Comment (valeria-meli) [reply javascript/ql/src/experimental/Security/CWE-918/SSRF.qll:19]: You're right. Sorry, I mixed it up with a similar query for Go (we did a more significant change there and I though we'd done that for Javascript too). Now I remember, we _tried_ to remove that heuristic but couldn't make it work.
We're testing your suggestion now :)

**Comment by asgerf** [comment]
@valeria-meli could you run the auto-formatter one more time? That's the only failing check. Otherwise LGTM.

**Review by asgerf — APPROVED** [review]
Thanks! 👍

**Pull request merged**